### PR TITLE
Forcibly reduce credits_observed to the last epoch

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

The initial inflation reward could be unexpected if inflation has not been enabled since genesis. (i.e. Virtually all of live clusters)
Subsequent inflation rewards are correct, though.

Conceptually, inflation rewards should be a _fair_ distribution of a stake-weighted contribution metrics (called points) across validators for some reasonable recent timespan.

The current implementation fulfils it for all inflation rewards but the first one.
The reason is that the initial inflation reward calculation is based on the full timespan since genesis.
Specifically, the initial inflation calculation uses all epochs since genesis as the timespan for points summation.

This is due to implantation remnant before periodic forced inflation rewards distribution (the `solana redeem` age) 
Technically, this is caused because stake account's `credits_observed` (key component of points calculation) aren't updated at all, until the very first inflation reward.
After that, as `credits_observed` is initialized, the inflation is calculated for each last epoch.

This creates the following counter intuitive inflation rewards.
As an extreme example, we reward _currently undelegated_ stake account. (But this happens only once at the first epoch after inflation is enabled)

Currently, `HDAhLXUZmFyvxT2gnGv36opsU6ZAXEEmcEzHMAghTBoT` is undelegated:

```
$ ./target/release/solana stake-account HDAhLXUZmFyvxT2gnGv36opsU6ZAXEEmcEzHMAghTBoT
Balance: 0.00228288 SOL
Rent Exempt Reserve: 0.00228288 SOL
Stake account is undelegated
Stake Authority: 9LQhe3fHy9i7RVWMiFiyon3GfC9rCzoXiwc9DjaAY3b7
Withdraw Authority: 9LQhe3fHy9i7RVWMiFiyon3GfC9rCzoXiwc9DjaAY3b7
```

But, if we would enable the inflation now, the stake account receives `+0.000016081` SOL. That's because initial inflation doesn't only look at the last epoch, but the entire blockchain history. So, this matches to the fact the stake account was actually delegating some SOLs (=~ ◎60) from epoch 58 to epoch 65:

```
        HDAhLXUZmFyvxT2gnGv36opsU6ZAXEEmcEzHMAghTBoT: Delegation {
            voter_pubkey: FKsC411dik9ktS6xPADxs4Fk2SCENvAiuccQHLAPndvk,
            stake: 59987617120,
            activation_epoch: 58,
            deactivation_epoch: 65,
            warmup_cooldown_rate: 0.25,
        },

```

```
HDAhLXUZmFyvxT2gnGv36opsU6ZAXEEmcEzHMAghTBoT(Stake11111111111111111111111111111111111111): ◎0.002282880 => ◎0.002298961 (+◎0.000016081 0.704417227%)
```

For comparison, similarly-sized stake account only receives `0.000000014` SOL.

```
H6rtteTo41ifz9XwMRx3yHXywssiGQizZYNCDQr4wZj1(Stake11111111111111111111111111111111111111): ◎0.010000000 => ◎0.010000014 (+◎0.000000014 0.000140000%)
```

In other words, staker will be rewarded proportional to the full delegated duration, not only the last epoch, with current implementation. So, this is a situation of act sooner and earn more.

#### Summary of Changes

Ideally, I think we should only consider the last epoch?
To realize that, this pr forcibly overwrites credits_observed before calculating inflation.

But this phenomenon is only one-time. and pico-inflation reward is very small. So, might not worth to fix this...

#### Context

Found via `ledger-tool capitalization` check.

### CODE REVIEW

**Comment by mvines** [comment]
> But this phenomenon is only one-time. and pico-inflation reward is very small. So, might not worth to fix this...

Yep we want to fix this!   The effect is small but we don't need a muddled story for how inflation gets turned on and having to explain why that first epoch after enabling inflation is special.

**Comment by ryoqun** [comment]
Wow, test passed.

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -1333,7 +1333,12 @@ impl Bank {
                     self.get_account(&stake_pubkey),
                     self.get_account(&delegation.voter_pubkey),
                 ) {
-                    (Some(stake_account), Some(vote_account)) => {
+                    (Some(mut stake_account), Some(vote_account)) => {
```
Comment (ryoqun) [inline runtime/src/bank.rs:1336]: this is least intrusive way. but a hacky way at the same time.

I'll do more commenting

**Comment by mvines** [comment]
This'll need a feature gate so it can be deployed out on testnet/devnet right?

**Comment by ryoqun** [comment]
> This'll need a feature gate so it can be deployed out on testnet/devnet right?

Yeah, it seems so. I originally thought that this shouldn't cause any calculation changes for cluster where inflation is already enabled. But it seems not. ;) I'm now working on this.

**Comment by mvines** [comment]
> > This'll need a feature gate so it can be deployed out on testnet/devnet right?
> 
> Yeah, it seems so. I originally thought that this shouldn't cause any calculation changes for cluster where inflation is already enabled. But it seems not. ;) I'm now working on this.

We turned off inflation for devnet/testnet to test the inflation kill switch, and to make it easier to deploy this across all clusters.  No gating required now I think!

File: programs/stake/src/stake_state.rs [hunk 2]
```diff
@@ -418,8 +418,8 @@ impl Stake {
         let mut points = 0;
         let mut new_credits_observed = self.credits_observed;
 
-        for (epoch, final_epoch_credits, initial_epoch_credits) in
-            new_vote_state.epoch_credits().iter().copied()
+        if let Some((epoch, final_epoch_credits, initial_epoch_credits)) =
+            new_vote_state.epoch_credits().last().copied()
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:422]: this is known that this ins't enough.

We need explicitly check whether the last epoch_credits is target_epoch for this inflation reward distribution. Otherwise,  we could be tricked by stale vote accounts.

File: programs/stake/src/stake_state.rs [hunk 3]
```diff
@@ -456,21 +457,39 @@ impl Stake {
     pub fn calculate_points_and_credits(
         &self,
         new_vote_state: &VoteState,
-        stake_history: Option<&StakeHistory>,
+        stake_context: Option<(Epoch, &StakeHistory)>,
         inflation_point_calc_tracer: &mut Option<impl FnMut(&InflationPointCalculationEvent)>,
     ) -> (u128, u64) {
         // if there is no newer credits since observed, return no point
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:463]: move comment along...

File: programs/stake/src/stake_state.rs (hunk 2 above)
Comment (ryoqun) [reply programs/stake/src/stake_state.rs:482]: now this is pushed.

**Comment by ryoqun** [comment]
- [ ] Also, add comment to note that now that VoteState.epoch_credits can be retired?

File: programs/stake/src/stake_state.rs [hunk 4]
```diff
@@ -456,21 +457,39 @@ impl Stake {
...
-        for (epoch, final_epoch_credits, initial_epoch_credits) in
-            new_vote_state.epoch_credits().iter().copied()
+        if let Some((epoch, final_epoch_credits, initial_epoch_credits)) =
+            new_vote_state.epoch_credits().last().copied()
         {
-            let stake = u128::from(self.delegation.stake(epoch, stake_history));
+            let stake = if let Some((target_epoch, _stake_history)) = stake_context {
+                if epoch != target_epoch {
+                    return (0, 0);
+                }
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:487]: write a test for this; this bug is later found after manual testing.

**Comment by t-nelson** [comment]
I happened back across this bug while investigating the viability of allowing fully-activated stake accounts to be merged.  Would an alternate (and cleaner?) solution be to ensure that the stake account's `observed_credits` field is updated even when there were no rewards to be paid out, regardless of split?  Caveat being that it needs feature gated, but stake-v2 isn't enabled yet :tada: 

I think this can be achieved by changing the early return here...
https://github.com/solana-labs/solana/blob/f25c969ad87e64e6d1fd07d2d37096ac71cf8d06/programs/stake/src/stake_state.rs#L598-L601
... to something like ...
```
let (voter_rewards, staker_rewards, is_split) = if rewards == 0 {
    (0, 0, false)
} else {
    vote_state.commission_split(rewards)
};
```
... then everything resolves itself at the next epoch boundary.

I say, "something like," because I haven't yet checked whether `rewards` is accounting for a sub-lamport situation at this here.

This solution has the side-effect that I believe we can safely merge two qualified fully activated stake accounts with just a few extra checks from the deactivated case.

@ryoqun @CriesofCarrots @mvines WDYT?

File: programs/stake/src/stake_state.rs [hunk 5]
```diff
@@ -507,26 +508,42 @@ impl Stake {
...
+        }
+
         if new_vote_state.credits() <= self.credits_observed {
             return (0, 0);
         }
 
         let mut points = 0;
         let mut new_credits_observed = self.credits_observed;
 
-        for (epoch, final_epoch_credits, initial_epoch_credits) in
```
Comment (t-nelson) [inline programs/stake/src/stake_state.rs:522]: Looking into this further... I believe this change can deny small stakes delegated to validators with small, but non-zero commissions ever getting paid out, as per https://github.com/solana-labs/solana/blob/d9a07e7ed91c9b67c30338b24f48e866cb9bae91/programs/stake/src/stake_state.rs#L611-L616

**Comment by CriesofCarrots** [comment]
> I think this can be achieved by changing the early return here...
> 
> https://github.com/solana-labs/solana/blob/f25c969ad87e64e6d1fd07d2d37096ac71cf8d06/programs/stake/src/stake_state.rs#L598-L601
> 
> ... to something like ...
> 
> ```
> let (voter_rewards, staker_rewards, is_split) = if rewards == 0 {
>     (0, 0, false)
> } else {
>     vote_state.commission_split(rewards)
> };
> ```
> 
> ... then everything resolves itself at the next epoch boundary.

@t-nelson , it looks to me like this will resolve the issue motivating this PR nicely. The only caveat, as we discussed in DM, is that we need to activate your fix at least one epoch before we activate inflation (to prevent the additional rewards that @ryoqun describes on the epoch that `observed_credits` gets bumped).
I will hold off activating stake-program-v2 on testnet so that we don't have to plumb an additional feature.

File: programs/stake/src/stake_state.rs (hunk 5 above)
Comment (ryoqun) [reply programs/stake/src/stake_state.rs:522]: yeah! correct! I intentionally done this way now. My two cent is posted here: https://github.com/solana-labs/solana/pull/13680#discussion_r527156860

File: programs/stake/src/stake_state.rs (hunk 4 above)
Comment (ryoqun) [reply programs/stake/src/stake_state.rs:541]: well, if we not exclude historical vote_state's epoch credits, this way, we might accidentally reward stale delegations well past its (last and stale) epoch_credits.

**Comment by ryoqun** [comment]
hehe. what a baroque work compared to the state of art pr #13680 by @t-nelson.

Closing this, because this has happily superseded. :)
